rust/src/detect.rs

```diff
@@ -0,0 +1,421 @@
+/* Copyright (C) 2022 Open Information Security Foundation
+ *
+ * You can copy, redistribute or modify this Program under the terms of
+ * the GNU General Public License version 2 as published by the Free
+ * Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * version 2 along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
+ * 02110-1301, USA.
+ */
+
+use nom7::branch::alt;
+use nom7::bytes::complete::{is_a, tag};
+use nom7::character::complete::digit1;
+use nom7::combinator::{complete, map_opt, opt, verify};
+use nom7::IResult;
+
+#[derive(PartialEq, Debug)]
+pub enum DetectUintMode {
+    DetectUintModeEqual,
+    DetectUintModeLt,
+    DetectUintModeGt,
+    DetectUintModeRange,
+    DetectUintModeNe,
+}
+
+#[derive(Debug)]
+pub struct DetectU32Data {
```

Comment:
@jasonish how do you recommend doing this ?
We will need the same structure and code for u8, u16, u32 and u64

Reply:
I guess one could use an enum?

```
enum UintValue {
    U8(u8),
    U16(u6),
    U32(u32),
}

struct DetectUintData {
    pub val: UintValue,
    pub valrange: UintValue,
}
```
it will at least reduce non-common code that is only size specific.

Reply:
With this, `val` could be u8 and `valrange` u16 ...
I will look into it a bit more, see how other crates handled it...

Reply:
I see that num-integer (used by suricata) uses macros for such as case cf https://github.com/rust-num/num-integer/blob/master/src/roots.rs#L165

@jasonish is not there a derive thing like `suricata_derive::AppLayerEvent` ?

Or maybe use crate `[num](https://docs.rs/num/0.4.0/num/index.html)::[integer](https://docs.rs/num/0.4.0/num/integer/index.html)::[Integer](https://github.com/OISF/suricata/pull/7117)`?

Reply:
> I see that num-integer (used by suricata) uses macros for such as case cf https://github.com/rust-num/num-integer/blob/master/src/roots.rs#L165
> 
> @jasonish is not there a derive thing like `suricata_derive::AppLayerEvent` ?

No, we'd have to write that.

> 
> Or maybe use crate `https://docs.rs/num/0.4.0/num/integer/trait.Integer.html` ?

If these fit your needs, then I think they are good to use. We already have them in our deps.  I don't think I've used them myself though.

Reply:
Managing something with rust generics... let's talk in next version of PR ;-)